// @repobit/dex-data-layer/dist/src/events/window-load-started-event/index.js
import { getCurrentDate, getCurrentTime } from '../utils.js';

class WindowLoadStartedEvent {
    event = 'page load started';
    pageInstanceID;
    page;
    constructor(pageInfo, pageAttributes = {}) {
        const existingPageLoadedEvent = window.adobeDataLayer.find(adobeDataLayerEntry => adobeDataLayerEntry.event === 'page load started');
        this.pageInstanceID = existingPageLoadedEvent.pageInstanceID;
        const existingPageLoadedObject = existingPageLoadedEvent.page;
        const info = pageInfo instanceof Function
            ? pageInfo({ ...existingPageLoadedObject.info })
            : {
                ...existingPageLoadedObject.info,
                ...pageInfo
            };
        this.page = {
            info,
            attributes: {
                ...existingPageLoadedObject.attributes,
                ...pageAttributes,
                time: getCurrentTime(),
                date: getCurrentDate()
            }
        };
    }
}

export { WindowLoadStartedEvent };
