// @repobit/dex-data-layer/dist/src/events/page-load-started-event/index.js
import '../../../../../dex-utils/dist/src/index.js';
import { getCurrentDate, getCurrentTime } from '../utils.js';
import UserAgent from '../../../../../dex-utils/dist/src/user-agent/index.js';

class PageLoadStartedEvent {
    event = 'page load started';
    pageInstanceID;
    page;
    constructor(page, pageInfo, pageAttributes = {}) {
        this.pageInstanceID = page.environment;
        const [section = '', subSection = '', subSubSection = '', subSubSubSection = ''] = pageInfo.name.split(':');
        const topLevelDomain = window.location.hostname.split('.').splice(1, 2)
            .join('.');
        this.page = {
            info: {
                ...pageInfo,
                section,
                subSection,
                subSubSection,
                subSubSubSection,
                destinationURL: window.location.href,
                queryString: window.location.search,
                referringURL: page.getParamValue('ref') || page.getParamValue('adobe_mc_ref') || document.referrer,
                serverName: window.location.hostname,
                sysEnv: UserAgent.os || '',
                language: page.locale
            },
            attributes: {
                promotionID: page.getParamValue('pid') || '',
                internalPromotionID: page.getParamValue('icid') || '',
                trackingID: page.getParamValue('cid') || '',
                time: getCurrentTime(),
                date: getCurrentDate(),
                domain: topLevelDomain,
                domainPeriod: window.location.hostname.split('.').length - 1,
                ...pageAttributes
            }
        };
    }
}

export { PageLoadStartedEvent };
